### ISSUE TITLE
isScopeRequired() false positive

### ISSUE THREAD

**Issue by b1rdex** [opis/closure#52]
Test case:
```php
function () {
    static $i = 0;
    return $i;
}
```

This `static` is for variable and doesn't reference the scope class.

It looks like this case should be removed:
https://github.com/opis/closure/blob/c37b40d59d9afffc2e5e958e2e5ccdb08634f81d/src/ReflectionClosure.php#L361-L364

**Comment by msarca** [comment]
Actually, that line is intentionally there. Normally we should call `$closure->bind($that, $scope)` every time we unserialize a closure, but that implies serializing `$this` object too, even if it's not used inside the closure. That would be painfully slow, so we try to detect first if `static`, `self`, `parent` or `$this` are used inside the closure's body. We do this by using a heuristic algorithm. So having a false positive from time to time is a good trade, especially since the result of that false positive is something the we should normally do (in this case, calling the `bind` method).

**Comment by mnapoli** [comment]
Thank you for your response @msarca.

I'm not sure I understand, maybe you can help: if there is the `static` keyword in the closure, it is considered to be using the current scope.

But `static $i` does not reference the current scope?

Why make that heuristic? I feel like I am missing something.

**Comment by msarca** [comment]
The *current scope* is a term that makes sense only when we are referring to the execution context. The class scope of a closure is something that is assigned to a closure when it's constructed, and once a closure was bound to a class scope it remains unchanged during the closure's lifetime, regardless of the execution context in which the closure runs.

```php
class A {
    static function test() {
        return "abc";
    }
    static function getClosure() {
        return function() {
            return static::test();
        };      
    }
}

class B {
    static function test() {
        return "xyz";
    }
    static function execute() {
        $closure = A::getClosure();
        // the current scope of
        // the execution context is irrelevant
        // it prints "abc", not "xyz"
        echo $closure();
    }
}

B::execute();
``` 
During the unserialization process, the reconstructed closure gets bound to the `Opis\Closure\SerializableClosure` class scope. Of course, this is not the closure's original class scope, so normally we should create a copy of the closure and bind it to the original class scope. But if `static`, `self` or `parent` are not used inside the closure, we can leave it as it is, because the closure's behavior will not be affected at all when executed. The fact that we're binding a closure to its original class scope when we encounter `static $i` is not wrong, it's just useless.

People tend to forget some simple and basic thing about PHP: if it runs, then it's syntactically correct. When serialization occurs we already know that there are no syntax errors, therefor we don't need to be redundant and use a full parser that will build an AST that was already built and make a syntax check which was already made. We just need to extract the code, and a heuristic approach is an optimal solution to this problem.

**Comment by mnapoli** [comment]
> The fact that we're binding a closure to its original class scope when we encounter `static $i` is not wrong, it's just useless.

I has some downsides though. We are trying to move PHP-DI from SuperClosure to opis/closure in https://github.com/PHP-DI/PHP-DI/pull/716.

The last blocker is this. Let me explain:

- if a closure references the scope, we throw an error: this is not allowed in PHP-DI config (we don't want to support that)
- however, closures with `static $i` are allowed

Since opis/closure says that `static $i` references the scope (which is not the case), we throw an exception when we shouldn't.

Is the problem more clear? This is why we would love to find a solution for `static $i`.

**Comment by sorinsarca** [comment]
@mnapoli this false-positive can be solved with a few lines of code on our side, but it will not solve all your issues:

- some closures may be internal - meaning that you can't see the code, can't see if they reference variables, and so on
- some closures may be constructed from callables - again, limited info
- the current approach (searching for $this, static, self, ... in closure's code) is not something that we want to rely on, so in the future versions if you have an object bound we will include the object in the serialization even if inside closure's code there is no $this, parent or similar. 

So, if you don't want to support bound closures, just put that in the docs and in code remove all bindings before serialization: `$closure = $closure->bindTo(null, null);`.

**Comment by msarca** [comment]
@mnapoli We had a small bug related to how we handle `null` class scope, but that was fixed in v3.5.3. Also you should be aware that unfortunately superclosure does not handle class scope correctly. 

```php
use Opis\Closure\ReflectionClosure;
use function Opis\Closure\{serialize as s, unserialize as u};
use SuperClosure\{SerializableClosure, Serializer, Analyzer\TokenAnalyzer};


class A {
    static function test() {
        echo "ok", PHP_EOL;
    }

    static function getClosure() {
        return function () {
            static::test();
        };
    }
}

$serializer = new Serializer();

$original = A::getClosure();

$opisclosure = u(s($original));
$superclosure = $serializer->unserialize($serializer->serialize($original));

var_dump("original", (new ReflectionFunction($original))->getClosureScopeClass());
var_dump("opisclosure", (new ReflectionFunction($opisclosure))->getClosureScopeClass());
var_dump("superclosure", (new ReflectionFunction($superclosure))->getClosureScopeClass());

echo "Run closure", PHP_EOL;

$original(); 
$opisclosure(); 
$superclosure(); 
```
Output 
```
string(8) "original"
object(ReflectionClass)#5 (1) {
  ["name"]=>
  string(1) "A"
}
string(11) "opisclosure"
object(ReflectionClass)#1138 (1) {
  ["name"]=>
  string(1) "A"
}
string(12) "superclosure"
NULL
Run closure
ok
ok

Fatal error: Uncaught Error: Cannot access static:: when no class scope is active
```

**Comment by mnapoli** [comment]
I am sorry but I'm not sure that we understand each other. I'll try to summarize it, let's start from scratch:

- PHP-DI does not support bound closures: we throw exceptions, and that's good for us (there's nothing to change here)
- a closure that uses `static $i` (NOT `static::XXX`) is, AFAIK, not using the scope

The problem is with `static $i` (static variables): we want to support those closures. And so we are at the root of the problem:

`static $i` is considered as "using the scope" by opis/closure, whereas it's not.

So to sum up: can we make opis/closure consider that `static $i` is not using the scope.

---

By the way, to clear any confusion, I am fine with calling `$closure->bindTo(null, null)` or doing anything necessary (since we don't care about closures that are scoped: these are not supported). It's just that `$closure->bindTo(null, null)` will not solve `static $i`.

**Comment by sorinsarca** [comment]
> By the way, to clear any confusion, I am fine with calling $closure->bindTo(null, null) or doing anything necessary (since we don't care about closures that are scoped: these are not supported). It's just that $closure->bindTo(null, null) will not solve static $i.

If you don't care about bound closures why are you still checking them? Just put in docs that bound closures are not supported and are automatically unbound at serialization.

Anyway, we already have a fix for this specific issue, a new version is created now.

We'll probably drop the `isScopeRequired` and `isBindingRequired` in the future major versions since there are a lot of cases that nobody can handle, so a lot of false-positive results may occur. That being said, `isScopeRequired()` will be `getClosureScopeClass() != null` and  `isBindingRequired()` will be `getClosureThis() != null`


```php
<?php
// include.php

echo $this->myProp;
```

```php
<?php

$f = function () {
  include "./include.php";
};

$f(); // throws
$f->bindTo((object)["myProp" => "works"]); // works

```

As you can see the closure is not bound but it requires a binding to work, and there is no way to know that, making `isScopeRequired` and `isBindingRequired` only hints instead of facts.

**Comment by msarca** [comment]
> So to sum up: can we make opis/closure consider that static $i is not using the scope.

@mnapoli We have fixed the issue with `static $i` and we'll tag a new version. 

> a closure that uses static $i (NOT static::XXX) is, AFAIK, not using the scope

Actually, that's not true and it can be easily proven.

```php
class A {
    static function getClosure() {
        return function() {
            static $i = 0;
        };
    }
}

$reflection = new ReflectionFunction(A::getClosure());
// Class scope not null
var_dump($reflection->getClosureScopeClass());

$f = (function() {
    static $i = 0;
})->bindTo(null, A::class);

$reflection = new ReflectionFunction($f);
// Class scope not null again
var_dump($reflection->getClosureScopeClass());
```
I'm just curious: why do you throw exceptions for bound closures and how do you currently detect if a closure was bound to a scope? Because you can not detect reliably if a closure is bound to a class scope or not when parsing (see the above example).

**Comment by mnapoli** [comment]
That's awesome if a fix is on the way in the next version, thanks we'll try that.

> >  a closure that uses static $i (NOT static::XXX) is, AFAIK, not using the scope
>
> Actually that's not true and it can be easily proven.

Well… yeah, you made that closure use a scope, so of course it _is using the scope_ 😉  I thought you would understand that I was implying that `static $i` does not use the scope "_unless something else in the closure uses the scope_". Anyway, I think we agree on things here.

**Comment by b1rdex** [comment]
Glad you guys came to the solution of the problem.
